cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadDumpAction.java

```diff
@@ -0,0 +1,71 @@
+package org.jenkinsci.plugins.workflow.cps;
+
+import hudson.Extension;
+import hudson.model.Action;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.concurrent.ExecutionException;
+import jenkins.model.TransientActionFactory;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+
+/**
+ * Shows thread dump for {@link CpsFlowExecution}.
+ */
+public final class CpsThreadDumpAction implements Action {
+
+    private final CpsFlowExecution execution;
+
+    private CpsThreadDumpAction(CpsFlowExecution execution) {
+        this.execution = execution;
+    }
+
+    @Override
+    public String getIconFileName() {
+        return "gear.png";
+    }
+
+    @Override
+    public String getDisplayName() {
+        return "Thread Dump";
```

Comment:
The current code lets anyone who has read permission on the build see the thread dump. I do not see any particular reason to restrict it; the steps being run are already visible in the build log, so the only additional information you would get is the function names and line numbers in scripts, which does not seem like it would be secret information. Do you feel otherwise?